tests/ops/test_stats.py

```diff
@@ -1,12 +1,90 @@
 import pytest
 import torch
 
-from pyro.ops.stats import (autocorrelation, autocovariance, effective_sample_size,
-                            gelman_rubin, split_gelman_rubin,
-                            _cummin, _fft_next_good_size)
+from pyro.ops.stats import (autocorrelation, autocovariance, effective_sample_size, gelman_rubin,
+                            hpdi, pi, quantile, resample, split_gelman_rubin, waic, _cummin,
+                            _fft_next_good_size)
 from tests.common import assert_equal, xfail_if_not_implemented
 
 
+@pytest.mark.parametrize('replacement', [True, False])
+@pytest.mark.init(rng_seed=3)
+def test_resample(replacement):
+    x = torch.empty(10000, 2)
+    x[:, 0].normal_(3, 4)
+    x[:, 1].normal_(5, 6)
+
+    num_samples = 5000
+    y = resample(x, num_samples=num_samples, replacement=replacement)
+    z = resample(x.t(), num_samples=num_samples, dim=1, replacement=replacement)
+    if not replacement:
+        assert_equal(torch.unique(y.reshape(-1)).numel(), y.numel())
+        assert_equal(torch.unique(z.reshape(-1)).numel(), z.numel())
+    assert_equal(y.shape, torch.Size([num_samples, 2]))
+    assert_equal(z.shape, torch.Size([2, num_samples]))
+    assert_equal(y.mean(dim=0), torch.tensor([3., 5.]), prec=0.1)
+    assert_equal(z.mean(dim=1), torch.tensor([3., 5.]), prec=0.12)
+    assert_equal(y.std(dim=0), torch.tensor([4., 6.]), prec=0.1)
+    assert_equal(z.std(dim=1), torch.tensor([4., 6.]), prec=0.1)
+
+
+@pytest.mark.init(rng_seed=3)
+def test_quantile():
+    x = torch.tensor([0., 1., 2.])
+    y = torch.rand(2000)
+    z = torch.randn(2000)
+
+    assert_equal(quantile(x, probs=[0., 0.4, 0.5, 1.]), torch.tensor([0., 0.8, 1., 2.]))
+    assert_equal(quantile(y, probs=0.2), torch.tensor(0.2), prec=0.01)
+    assert_equal(quantile(z, probs=0.8413), torch.tensor(1.), prec=0.001)
+
+
+def test_pi():
+    x = torch.empty(1000).log_normal_(0, 1)
+    assert_equal(pi(x, prob=0.8), quantile(x, probs=[0.1, 0.9]))
+
+
+@pytest.mark.init(rng_seed=3)
+def test_hpdi():
+    x = torch.randn(20000)
+    assert_equal(hpdi(x, prob=0.8), pi(x, prob=0.8), prec=0.01)
+
+    x = torch.empty(20000).exponential_(1)
+    assert_equal(hpdi(x, prob=0.2), torch.tensor([0., 0.22]), prec=0.01)
+
+
+def _quantile(x, dim=0):
+    return quantile(x, probs=[0.1, 0.6], dim=dim)
+
+
+def _pi(x, dim=0):
+    return pi(x, prob=0.8, dim=dim)
+
+
+def _hpdi(x, dim=0):
+    return hpdi(x, prob=0.8, dim=dim)
+
+
+@pytest.mark.parametrize('statistics', [_quantile, _pi, _hpdi])
+@pytest.mark.parametrize('sample_shape', [(), (3,), (2, 3)])
```

Comment:
Any reason why we can't add these to the parameters below?

Reply:
These statistics generate results with shapes changed. The `autocovariance` e.g. gives the same shape.